src/lib/HasuraClient.ts

```diff
@@ -11,13 +12,32 @@ const authLink = setContext((_, { headers }) => {
   return {
     headers: {
       ...headers,
-      authorization: token ? `Bearer ${token}` : "",
+      /* 
+        We send a value here ("Bearer ") even when token is 
+        missing because Hasura interprets missing authorization 
+        headers and even empty string as a public user.
+
+        Instead we want Hasura to return an "invalid-jwt" error
+        so that we can check for it as a condition to refresh
+        the token.
+      */
+      authorization: `Bearer ${token}`,
     },
   };
 });
 
+const errorLink = onError(({ graphQLErrors }) => {
+  if (graphQLErrors) {
+    graphQLErrors.map(({ extensions }) => {
+      if (extensions?.code === "invalid-jwt") {
+        AuthenticationService.refreshHasuraToken();
+      }
+    });
+  }
```

Comment:
- You're using a `map` without actually mapping to anything, `forEach` is what you want here
- You're not awaiting `refreshHasuraToken()`

(note, you won't be able to use either map or forEach with an asynchronous operation in it

Suggestion:

```suggestion
const invalidToken = (graphQLErrors || []).find(({ extensions }) => {
  return extensions?.code === "invalid-jwt"
})

if (invalidToken) await AuthenticationService.refreshHasuraToken();
```

Also note how to retry the graphql query here: https://www.apollographql.com/docs/react/data/error-handling/#retrying-operations

Reply:
Updated. As for the retry logic I'm implementing it in a subsequent PR. Can't use `await` inside `onError` because it doesn't accept promises, so the example given in the link doesn't work here.

Reply:
Then gotta find something else. This will just cause the retry to fail again and again cause it won't wait for the token refresh

Reply:
you can change the `getHasuraToken()` so that it fetches the token if absent.
And onError can do `AuthenticationService.invalidateHasuraToken()`

Reply:
But we're not merging this with an async function flying solo

Reply:
Or just go straight for the RetryLink https://www.apollographql.com/docs/react/api/link/apollo-link-retry/